### PR TITLE
Blockchain module cleanup

### PR DESCRIPTION
**Pull request by debris**
changes:

- split `block_to_extras_update` function into 6 smaller functions. Code is cleaner, more maintainable and readable.
- fully separated preparing extas update from applying it to a batch. 
- split `blockchain.rs` into few separate files. This file was becoming a behemoth.
- fixed cleaning up blockchain caches.

### CODE REVIEW

**Comment by debris** [comment]
I recommend opening `blockchain.rs` file in new tab. Git diff in this case is hardly readable.

**Comment by gavofyork** [comment]
needs a rebase. also it's quite big - i presume no logic has been added in this PR?

**Comment by debris** [comment]
no, I will rebase

File: ethcore/src/blockchain/blockchain.rs [hunk 1]
```diff
@@ -452,40 +355,26 @@ impl BlockChain {
...
-	fn tree_route_aux(&self, from: (&BlockDetails, &H256), to: (&BlockDetails, &H256)) -> TreeRoute {
+	pub fn tree_route(&self, from: H256, to: H256) -> TreeRoute {
 		let mut from_branch = vec![];
 		let mut to_branch = vec![];
 
-		let mut from_details = from.0.clone();
-		let mut to_details = to.0.clone();
-		let mut current_from = from.1.clone();
-		let mut current_to = to.1.clone();
+		let mut from_details = self.block_details(&from).expect(&format!("0. Expected to find details for block {:?}", from));
```
Comment (gavofyork) [inline ethcore/src/blockchain/blockchain.rs]: this probably shouldn't panic.

File: ethcore/src/blockchain/blockchain.rs [hunk 2]
```diff
@@ -452,40 +355,26 @@ impl BlockChain {
...
+		let mut to_details = self.block_details(&to).expect(&format!("1. Expected to find details for block {:?}", to));
```
Comment (gavofyork) [inline ethcore/src/blockchain/blockchain.rs]: this probably shouldn't panic.

File: ethcore/src/blockchain/blockchain.rs [hunk 3]
```diff
@@ -452,40 +355,26 @@ impl BlockChain {
...
+		let mut current_from = from;
+		let mut current_to = to;
 
 		// reset from && to to the same level
 		while from_details.number > to_details.number {
 			from_branch.push(current_from);
 			current_from = from_details.parent.clone();
-			from_details = self.block_details(&from_details.parent).expect(&format!("1. Expected to find details for block {:?}", from_details.parent));
+			from_details = self.block_details(&from_details.parent).expect(&format!("2. Expected to find details for block {:?}", from_details.parent));
```
Comment (gavofyork) [inline ethcore/src/blockchain/blockchain.rs]: this probably shouldn't panic.

File: ethcore/src/blockchain/blockchain.rs [hunk 4]
```diff
@@ -452,40 +355,26 @@ impl BlockChain {
...
 		}
 
 		while to_details.number > from_details.number {
 			to_branch.push(current_to);
 			current_to = to_details.parent.clone();
-			to_details = self.block_details(&to_details.parent).expect(&format!("2. Expected to find details for block {:?}", to_details.parent));
+			to_details = self.block_details(&to_details.parent).expect(&format!("3. Expected to find details for block {:?}", to_details.parent));
```
Comment (gavofyork) [inline ethcore/src/blockchain/blockchain.rs]: this probably shouldn't panic.

File: ethcore/src/blockchain/blockchain.rs [hunk 5]
```diff
@@ -494,11 +383,11 @@ impl BlockChain {
 		while current_from != current_to {
 			from_branch.push(current_from);
 			current_from = from_details.parent.clone();
-			from_details = self.block_details(&from_details.parent).expect(&format!("3. Expected to find details for block {:?}", from_details.parent));
+			from_details = self.block_details(&from_details.parent).expect(&format!("4. Expected to find details for block {:?}", from_details.parent));
```
Comment (gavofyork) [inline ethcore/src/blockchain/blockchain.rs]: this probably shouldn't panic.

File: ethcore/src/blockchain/blockchain.rs [hunk 6]
```diff
@@ -494,11 +383,11 @@ impl BlockChain {
...
 
 			to_branch.push(current_to);
 			current_to = to_details.parent.clone();
-			to_details = self.block_details(&to_details.parent).expect(&format!("4. Expected to find details for block {:?}", from_details.parent));
+			to_details = self.block_details(&to_details.parent).expect(&format!("5. Expected to find details for block {:?}", from_details.parent));
```
Comment (gavofyork) [inline ethcore/src/blockchain/blockchain.rs]: this probably shouldn't panic.

File: ethcore/src/blockchain/blockchain.rs [hunk 7]
```diff
@@ -527,165 +416,225 @@ impl BlockChain {
...
+		let parent_hash = header.parent_hash();
+		let parent_details = self.block_details(&parent_hash).expect(format!("Invalid parent hash: {:?}", parent_hash).as_ref());
 		let total_difficulty = parent_details.total_difficulty + header.difficulty();
 		let is_new_best = total_difficulty > self.best_block_total_difficulty();
+
+		BlockInfo {
+			hash: hash,
+			number: number,
+			total_difficulty: total_difficulty,
+			location: match is_new_best {
```
Comment (gavofyork) [inline ethcore/src/blockchain/blockchain.rs]: consider rewriting as:
```
if is_new_best {
	// on new best block we need to make sure that all ancestors
	// are moved to "canon chain"
	// find the route between old best block and the new one
	let best_hash = self.best_block_hash();
	let route = self.tree_route(best_hash, parent_hash);

	assert_eq!(number, parent_details.number + 1);

	match route.blocks.len() {
		0 => BlockLocation::CanonChain,
		_ => BlockLocation::BranchBecomingCanonChain {
			ancestor: route.ancestor,
			route: route.blocks.into_iter().skip(route.index).collect()
		}
	}
} else { BlockLocation::Branch }
```
(saves one line and avoids the `true` and `false` symbols).

File: ethcore/src/blockchain/blockchain.rs [hunk 8]
```diff
@@ -527,165 +416,225 @@ impl BlockChain {
...
+			})
+	}
 
-		match route.blocks.len() {
-			// its our parent
-			1 => { 
-				batch.put_extras(&header.number(), &hash);
+	/// This functions returns modified blocks blooms.
+	///
+	/// TODO: this function needs comprehensive description.
```
Comment (gavofyork) [inline ethcore/src/blockchain/blockchain.rs]: maybe do this :)

**Comment by gavofyork** [comment]
it's not especially clear that no logic has been added/changed. maybe add a commentary so we know which functions have been moved where?

given that the tests pass, i'm inclined to let it go this time, but i think for complex refactoring, when it's not clear that it is just repotting, either break down the PRs into two or more smaller refactorings or give clear commentary about what's going on in the PR (e.g. line comments seem to work quite well to guide the reviewer).

File: ethcore/src/blockchain/blockchain.rs (hunk 7 above)
Comment (debris) [inline ethcore/src/blockchain/blockchain.rs]: actually if statement requires more lines, cause I have to wrap it into additional `{ }`

eg.

```rust
BlockInfo {
  hash: hash,
  number: number,
  total_difficulty: total_difficulty,
  location: {
    if is_new_best {
      // ...
    } else {
      // ...
    }
  }
}
```

File: ethcore/src/blockchain/blockchain.rs (hunk 8 above)
Comment (debris) [inline ethcore/src/blockchain/blockchain.rs]: sure

File: ethcore/src/blockchain/blockchain.rs (hunk 1 above)
Comment (debris) [inline ethcore/src/blockchain/blockchain.rs]: Actually here it should. Missing block details means one of two things:

- our db is malformed (or not complete).
- this function is being used incorrectly.

**Comment by arkpar** [comment]
There should be a test that checks if the blockchain cache has been updated correctly after a fork

File: ethcore/src/blockchain/blockchain.rs (hunk 7 above)
Comment (gavofyork) [inline ethcore/src/blockchain/blockchain.rs]: i think not (your example isn't a fair comparison since you're placing the `if` on its own line whereas `match went on the `location:` line), and requiring the `else` branch to sit on its own line (i would say trivial `else` branches may reasonably be inlined). so now you've got:

```
match is_new_best {
  false => BlockLocation::Branch,
  true => {
    /* 15 lines */
  }
}
```

(20 lines)

versus

```
if is_new_best {
  /* 15 lines */
} else { BlockLocation::Branch }
```

(17 lines)

even if you place the `else` branch on its own line, it's still a line shorter:

```
if is_new_best {
  /* 15 lines */
} else {
  BlockLocation::Branch
}
```

(19 lines)

this way you have 1 fewer indentation levels (elided body is at 1 indent with `if` rather than 2 with `match`) and 4 fewer symbols in the code (`if` + `is_new_best` + `{` + `}` + `else` + `{` + `BlockLocation::Branch` + `}` = 8, versus `match` + `is_new_best` + `{` + `false` + `=>` + `BlockLocation::Branch` + `,` + `true` + `=>` + `{` + `}` + `}` = 12).

`if` wins.

File: ethcore/src/blockchain/blockchain.rs (hunk 1 above)
Comment (gavofyork) [inline ethcore/src/blockchain/blockchain.rs]: asserting on an external factor such as what is or is not in some particular portion of the database is not particularly defensive coding.

the only proper use of panics/assertions in live code is for specifying logical certainties (`assert!(1 != 0);`) or absolute assertions over the high-level design that are themselves well tested (e.g. triemutdb asserts the root is in the database).

it is sometimes reasonable to use them to specify postconditions when used in order to specify a function's preconditions, this in this case they should be simple enough that it can be done up front in the code prior to the implementation proper.

if you cannot point to comprehensive tests and the structure of the software's design that means your assertion _can never be_ false, then it should not panic, but rather fail-over at a higher level.

in this case, there is a mere "expectation" that the database happens to have a particular entry in it, derived from a potentially arbitrary other number of entries, themselves each populated by code in other modules, and all taken from an argument of unknown provenance.

in short, unless you are willing to bet the farm on this function _never_ being used "improperly", and can provide that by design it cannot happen, then a panic is unacceptable in this position.

**Comment by gavofyork** [comment]
can "fixed cleaning up blockchain caches" be placed in a separate PR for ease of review? the other 3 changes this PR makes change no logic.

**Comment by debris** [comment]
@gavofyork 
> it's not especially clear that no logic has been added/changed. maybe add a commentary so we know which functions have been moved where?

- moved `blockchain.rs` to `blockchain/blockchain.rs`
- added `BlockInfo` and `BlockLocation` structs (`blockchain/block_info.rs`). They replaced unreadable if statements and comments
- split `block_to_extras_update` into 6 functions. Each of this functions updates only one extras type.
  - `block_info`
  - `prepare_block_hashes_update`
  - `prepare_block_details_update`
  - `prepare_block_receipts_update`
  - `prepare_transaction_addresses_update`
  - `prepare_block_blooms_update`
- moved `BloomIndexer` and it's tests to `blockchain/bloom_indexer.rs`
- moved `BestBlock` to `blockchain/best_block.rs`
- moved `CacheSize` to `blockchain/cache.rs`
- moved `TreeRoute` to blockchain/tree_route.rs`
- moved `ExtrasUpdate` to blockchain/update.rs
- removed `tree_route_aux` function which was not needed (previously used in `block_to_extras_update` function)

@gavofyork 
> can "fixed cleaning up blockchain caches" be placed in a separate PR for ease of review? the other 3 changes this PR makes change no logic.

These are just two lines. [block hashes](https://github.com/ethcore/parity/blob/blockchain_cleanup/ethcore/src/blockchain/blockchain.rs#L445)
and [transaction addresses](https://github.com/ethcore/parity/blob/blockchain_cleanup/ethcore/src/blockchain/blockchain.rs#L463).

@arkpar
> There should be a test that checks if the blockchain cache has been updated correctly after a fork

There are already few tests which check updating block_details && blocks_blooms after a fork. I think it's best to add more test for forking in next prs, cause these tests will make this pr significantly bigger.

File: ethcore/src/blockchain/blockchain.rs (hunk 8 above)
Comment (debris) [inline ethcore/src/blockchain/blockchain.rs]: docs updated :)

File: ethcore/src/blockchain/blockchain.rs (hunk 1 above)
Comment (debris) [inline ethcore/src/blockchain/blockchain.rs]: I see. I will fix it in next prs.

File: ethcore/src/blockchain/blockchain.rs (hunk 7 above)
Comment (debris) [inline ethcore/src/blockchain/blockchain.rs]: > i think not (your example isn't a fair comparison since you're placing the if on its own line 

I thought that I have to write if inside bracers (like its required to do when if is on the right side of `=>`), but it wasn't necessary here... my bad. Fixed in latest commit.

**Comment by gavofyork** [comment]
ok to merge on my end, though i'd like to see the fork tests and removal of asserts in the not-too-distant future. perhaps introduce an issue to track?

**Comment by debris** [comment]
I've added #529 and #530

**Comment by debris** [comment]
Started working on sweeping panics on a separate branch. Does `tree_route` function look better? https://github.com/ethcore/parity/commit/8c61131b98a87e29bde585f78aafba98164c77f1#diff-422bab2793f9fb234587f1dadcf690d3R350
